File: TypeScript/clbg/n-body/n-body.js
const N = parseInt(process.argv[2] ?? "0", 10);
const PI = 3.141592653589793;
const SOLAR_MASS = 4.0 * PI * PI;
const DAYS_PER_YEAR = 365.24;
const px = new Float64Array(5);
const py = new Float64Array(5);
const pz = new Float64Array(5);
const vx = new Float64Array(5);
const vy = new Float64Array(5);
const vz = new Float64Array(5);
const mass = new Float64Array(5);
function offsetMomentum() {
    let pxm = 0.0, pym = 0.0, pzm = 0.0;
    for (let i = 0; i < 5; i++) {
        pxm += vx[i] * mass[i];
        pym += vy[i] * mass[i];
        pzm += vz[i] * mass[i];
    }
    vx[0] = -pxm / SOLAR_MASS;
    vy[0] = -pym / SOLAR_MASS;
    vz[0] = -pzm / SOLAR_MASS;
}
function initBodies() {
    // Sun
    px[0] = 0;
    py[0] = 0;
    pz[0] = 0;
    vx[0] = 0;
    vy[0] = 0;
    vz[0] = 0;
    mass[0] = SOLAR_MASS;
    // Jupiter
    px[1] = 4.84143144246472090e+00;
    py[1] = -1.16032004402742839e+00;
    pz[1] = -1.03622044471123109e-01;
    vx[1] = 1.66007664274403694e-03 * DAYS_PER_YEAR;
    vy[1] = 7.69901118419740425e-03 * DAYS_PER_YEAR;
    vz[1] = -6.90460016972063023e-05 * DAYS_PER_YEAR;
    mass[1] = 9.54791938424326609e-04 * SOLAR_MASS;
    // Saturn
    px[2] = 8.34336671824457987e+00;
    py[2] = 4.12479856412430479e+00;
    pz[2] = -4.03523417114321381e-01;
    vx[2] = -2.76742510726862411e-03 * DAYS_PER_YEAR;
    vy[2] = 4.99852801234917238e-03 * DAYS_PER_YEAR;
    vz[2] = 2.30417297573763929e-05 * DAYS_PER_YEAR;
    mass[2] = 2.85885980666130812e-04 * SOLAR_MASS;
    // Uranus
    px[3] = 1.28943695621391310e+01;
    py[3] = -1.51111514016986312e+01;
    pz[3] = -2.23307578892655734e-01;
    vx[3] = 2.96460137564761618e-03 * DAYS_PER_YEAR;
    vy[3] = 2.37847173959480950e-03 * DAYS_PER_YEAR;
    vz[3] = -2.96589568540237556e-05 * DAYS_PER_YEAR;
    mass[3] = 4.36624404335156298e-05 * SOLAR_MASS;
    // Neptune
    px[4] = 1.53796971148509165e+01;
    py[4] = -2.59193146099879641e+01;
    pz[4] = 1.79258772950371181e-01;
    vx[4] = 2.68067772490389322e-03 * DAYS_PER_YEAR;
    vy[4] = 1.62824170038242295e-03 * DAYS_PER_YEAR;
    vz[4] = -9.51592254519715870e-05 * DAYS_PER_YEAR;
    mass[4] = 5.15138902046611451e-05 * SOLAR_MASS;
    offsetMomentum();
}
function energy() {
    let e = 0.0;
    for (let i = 0; i < 5; i++) {
        const m = mass[i];
        e += 0.5 * m * (vx[i] * vx[i] + vy[i] * vy[i] + vz[i] * vz[i]);
        for (let j = i + 1; j < 5; j++) {
            const dx = px[i] - px[j];
            const dy = py[i] - py[j];
            const dz = pz[i] - pz[j];
            const dist = Math.sqrt(dx * dx + dy * dy + dz * dz);
            e -= (m * mass[j]) / dist;
        }
    }
    return e;
}
function advance(dt) {
    for (let i = 0; i < 5; i++) {
        for (let j = i + 1; j < 5; j++) {
            const dx = px[i] - px[j];
            const dy = py[i] - py[j];
            const dz = pz[i] - pz[j];
            const d2 = dx * dx + dy * dy + dz * dz;
            const dist = Math.sqrt(d2);
            const mag = dt / (d2 * dist);
            const mi = mass[i];
            const mj = mass[j];
            const s1 = mj * mag;
            const s2 = mi * mag;
            vx[i] -= dx * s1;
            vy[i] -= dy * s1;
            vz[i] -= dz * s1;
            vx[j] += dx * s2;
            vy[j] += dy * s2;
            vz[j] += dz * s2;
        }
    }
    for (let i = 0; i < 5; i++) {
        px[i] += dt * vx[i];
        py[i] += dt * vy[i];
        pz[i] += dt * vz[i];
    }
}
initBodies();
const initial = energy();
for (let i = 0; i < N; i++)
    advance(0.01);
const final = energy();
process.stdout.write(initial.toFixed(9) + "\n" + final.toFixed(9) + "\n");

File: Java/clbg/n-body/nbody.java
import java.util.Locale;

public class nbody {
    private static final double SOLAR_MASS = 4.0 * Math.PI * Math.PI;
    private static final double DAYS_PER_YEAR = 365.24;
    private static final double DT = 0.01;

    private static final class Body {
        double x, y, z;
        double vx, vy, vz;
        final double mass;

        Body(double x, double y, double z,
             double vx, double vy, double vz, double mass) {
            this.x = x;
            this.y = y;
            this.z = z;
            this.vx = vx * DAYS_PER_YEAR;
            this.vy = vy * DAYS_PER_YEAR;
            this.vz = vz * DAYS_PER_YEAR;
            this.mass = mass * SOLAR_MASS;
        }
    }

    private static Body[] createSystem() {
        return new Body[] {
            new Body(0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 1.0),
            new Body(
                4.84143144246472090,
                -1.16032004402742839,
                -1.03622044471123109e-1,
                1.66007664274403694e-3,
                7.69901118419740425e-3,
                -6.90460016972063023e-5,
                9.54791938424326609e-4
            ),
            new Body(
                8.34336671824457987,
                4.12479856412430479,
                -4.03523417114321381e-1,
                -2.76742510726862411e-3,
                4.99852801234917238e-3,
                2.30417297573763929e-5,
                2.85885980666130812e-4
            ),
            new Body(
                1.28943695621391310e1,
                -1.51111514016986312e1,
                -2.23307578892655734e-1,
                2.96460137564761618e-3,
                2.37847173959480950e-3,
                -2.96589568540237556e-5,
                4.36624404335156298e-5
            ),
            new Body(
                1.53796971148509165e1,
                -2.59193146099879641e1,
                1.79258772950371181e-1,
                2.68067772490389322e-3,
                1.62824170038242295e-3,
                -9.51592254519715870e-5,
                5.15138902046611451e-5
            )
        };
    }

    private static void offsetMomentum(Body[] bodies) {
        double px = 0.0;
        double py = 0.0;
        double pz = 0.0;

        for (Body b : bodies) {
            px += b.vx * b.mass;
            py += b.vy * b.mass;
            pz += b.vz * b.mass;
        }

        bodies[0].vx = -px / SOLAR_MASS;
        bodies[0].vy = -py / SOLAR_MASS;
        bodies[0].vz = -pz / SOLAR_MASS;
    }

    private static double energy(Body[] bodies) {
        double result = 0.0;

        for (int i = 0; i < bodies.length; i++) {
            Body a = bodies[i];
            result += 0.5 * a.mass
                    * (a.vx * a.vx + a.vy * a.vy + a.vz * a.vz);

            for (int j = i + 1; j < bodies.length; j++) {
                Body b = bodies[j];
                double dx = a.x - b.x;
                double dy = a.y - b.y;
                double dz = a.z - b.z;
                double distance = Math.sqrt(dx * dx + dy * dy + dz * dz);
                result -= (a.mass * b.mass) / distance;
            }
        }

        return result;
    }

    private static void advance(Body[] bodies, int steps) {
        final int count = bodies.length;

        for (int step = 0; step < steps; step++) {
            for (int i = 0; i < count - 1; i++) {
                Body a = bodies[i];
                double ax = a.x;
                double ay = a.y;
                double az = a.z;
                double avx = a.vx;
                double avy = a.vy;
                double avz = a.vz;
                double amass = a.mass;

                for (int j = i + 1; j < count; j++) {
                    Body b = bodies[j];
                    double dx = ax - b.x;
                    double dy = ay - b.y;
                    double dz = az - b.z;
                    double distanceSquared = dx * dx + dy * dy + dz * dz;
                    double magnitude = DT
                            / (distanceSquared * Math.sqrt(distanceSquared));

                    avx -= dx * b.mass * magnitude;
                    avy -= dy * b.mass * magnitude;
                    avz -= dz * b.mass * magnitude;

                    b.vx += dx * amass * magnitude;
                    b.vy += dy * amass * magnitude;
                    b.vz += dz * amass * magnitude;
                }

                a.vx = avx;
                a.vy = avy;
                a.vz = avz;
            }

            for (int i = 0; i < count; i++) {
                Body b = bodies[i];
                b.x += DT * b.vx;
                b.y += DT * b.vy;
                b.z += DT * b.vz;
            }
        }
    }

    public static void main(String[] args) {
        int n = Integer.parseInt(args[0]);
        Body[] bodies = createSystem();
        offsetMomentum(bodies);

        double initialEnergy = energy(bodies);
        advance(bodies, n);
        double finalEnergy = energy(bodies);

        System.out.printf(Locale.ROOT, "%.9f\n%.9f\n",
                          initialEnergy, finalEnergy);
    }
}